website_form_validations/controllers/main.py

```diff
@@ -0,0 +1,218 @@
+                    domain = [("partner_id", "=", order.partner_id.id)]
+                    user_ids = request.registry['res.users'].search(
+                        cr,
+                        SUPERUSER_ID,
+                        domain,
+                        context=dict(
+                            context or {},
+                            active_test=False))
+                    if not user_ids or request.website.user_id.id not in user_ids:  # noqa
+                        checkout.update(
+                            self.checkout_parse(
+                                "billing",
+                                order.partner_id))
+        else:
+            checkout = self.checkout_parse('billing', data)
+            try:
+                shipping_id = int(data["shipping_id"])
+            except ValueError:
+                pass
+            if shipping_id == -1:
+                checkout.update(self.checkout_parse('shipping', data))
+
+        if shipping_id is None:
+            if not order:
+                order = request.website.sale_get_order(context=context)
+            if order and order.partner_shipping_id:
+                shipping_id = order.partner_shipping_id.id
+
+        shipping_ids = list(set(shipping_ids) - set([partner.id]))
+
+        if shipping_id == partner.id:
+            shipping_id = 0
+        elif shipping_id > 0 and shipping_id not in shipping_ids:
+            shipping_ids.append(shipping_id)
+        elif shipping_id is None and shipping_ids:
+            shipping_id = shipping_ids[0]
+
+        ctx = dict(context, show_address=1)
+        shippings = []
+        if shipping_ids:
+            shippings = shipping_ids and orm_partner.browse(
+                cr,
+                SUPERUSER_ID,
+                list(shipping_ids),
+                ctx) or []
+        if shipping_id > 0:
+            shipping = orm_partner.browse(cr, SUPERUSER_ID, shipping_id, ctx)
+            checkout.update(self.checkout_parse("shipping", shipping))
+
+        checkout['shipping_id'] = shipping_id
+
+        # Default search by user country
+        if not checkout.get('country_id'):
+            country_code = request.session['geoip'].get('country_code')
+            if country_code:
+                country_ids = request.registry.get('res.country').search(
+                    cr, uid, [
+                        ('code', '=', country_code)], context=context)
+                if country_ids:
+                    checkout['country_id'] = country_ids[0]
+        checkout['mobile'] = partner.mobile
+        values = {
+            'countries': countries,
+            'states': states,
+            'checkout': checkout,
+            'shipping_id': partner.id != shipping_id and shipping_id or 0,
+            'shippings': shippings,
+            'error': {},
+            'has_check_vat': hasattr(registry['res.partner'], 'check_vat')
+        }
+
+        return values
+
+    def checkout_form_save(self, checkout):
```

Comment:
I cannot do "Too much bussiness logic" thing, this is a very wired feature that requires a lot of data processing, this data is retrieved from the website, and because of the nature of the inherited controller this is necessary.

On the other hand I can try to split the code in several lines, but cant promise it will be good.

on the "To much IF" thing you can see this is not complex code and I'm not violating code complexity rules: https://en.wikipedia.org/wiki/Cyclomatic_complexity 
this was also checked by McCabe Complexity checker on my local instance: https://pypi.python.org/pypi/mccabe